drivers/rail/rail_netdev.c

```diff
@@ -0,0 +1,496 @@
+                                         pack_handle,
+                                         &pack_info
+                                         );
+    dev->lastRxPacketHandle = pack_handle;
+
+    // buf == NULL && len == 0 -> return packet size, no dropping
+    if (buf == NULL && len == 0) {
+        DEBUG("_recv: no dropping return packet size: 0x%02x\n", pack_info.packetBytes);
+        // -1 because no len infos
+        return pack_info.packetBytes - 1;
+    }
+    // buf == NULL && len > 0 -> return packet size + drop it
+    if (buf == NULL && len > 0) {
+        // drop it
+        DEBUG("_recv: drop packet - return packet size: 0x%02x\n", pack_info.packetBytes);
+        RAIL_ReleaseRxPacket(dev->rhandle, pack_handle);
+        dev->lastRxPacketHandle = RAIL_RX_PACKET_HANDLE_INVALID;
+        // -1 because no len infos
+        return pack_info.packetBytes;
+    }
+
+    // clear info struct
+    memset(&pack_details, 0, sizeof(RAIL_RxPacketDetails_t));
+
+    // get more infos about the packet
+    ret = RAIL_GetRxPacketDetails(dev->rhandle, pack_handle, &pack_details);
+
+    if (ret != RAIL_STATUS_NO_ERROR) {
+        LOG_ERROR("Error receiving new packet / frame - msg %s\n", rail_error2str(ret));
+        return -1;
+    }
+
+    DEBUG("time received: %lu\n"
+          "crcStatus %s \n"
+          "isAck: %s\n"
+          "subPhy: %u\n"
+          "rssiLatch: %d dBm\n"
+          "lqi: %u\n"
+          "syncWordId: %u\n"
+          "antenna id: %u\n"
+          "payload size: %u\n",
+          pack_details.timeReceived.packetTime,
+          pack_details.crcPassed ? "Passed" : "Failed",
+          pack_details.isAck ? "Ack" : "Not a Ack",
+          pack_details.subPhyId,
+          pack_details.rssi,
+          pack_details.lqi,
+          pack_details.syncWordId,
+          pack_details.antennaId,
+          pack_info.packetBytes
+          );
+
+    // question: with length info in byte 0 or without?
+    // first try without, skip it
+    pack_info.firstPortionData++;
+    pack_info.firstPortionBytes--;
+    pack_info.packetBytes--;
+
+    // copy payload of packet
+    RAIL_CopyRxPacket((uint8_t *)buf, &pack_info);
+
+    /*
+       DEBUG("Print buf cpy size %d: ", cpy_size);
+       for (int i = 0; i < cpy_size; i++) {
+        if (i % 4 == 0) DEBUG("\n");
+        DEBUG("0x%02x ", ((uint8_t*)buf)[i]);
+       }
+       DEBUG("\n");
+     */
+    if (info != NULL) {
+        netdev_ieee802154_rx_info_t *rx_info = info;
+        rx_info->rssi = pack_details.rssi;
+        rx_info->lqi = pack_details.lqi;
+    }
+
+    // free packet, set handle to null
+    RAIL_ReleaseRxPacket(dev->rhandle, pack_handle);
+    dev->lastRxPacketHandle = RAIL_RX_PACKET_HANDLE_INVALID;
+
+    return pack_info.packetBytes;
+}
+
+static void _isr(netdev_t *netdev)
+{
+    DEBUG("rail_netdev->isr called\n");
+    // dunno what to do, but call the callback ...
+    netdev->event_callback(netdev, NETDEV_EVENT_RX_COMPLETE);
```

Comment:
Here you're suppose to call the event callback with the right parameter. For that, you need to parse the actual reason of the irq (irq mask or somehting like that).

Reply:
well this is not necessary, because the silabs driver blob already handles this. 
In some way the ``_rail_radio_event_handler`` in ``rail_drv.c`` is the semantic equivalent to the ``_isr()`` function. 

If I could, I would skip the ``_isr()`` and send the ``NETDEV_EVENT_RX_COMPLETE`` event directly, but this mixes something up in the stack and well it doesn't work (I tried it ...)

But I have to do most of the stuff in the ``_rail_radio_event_handler()``, because I can not attach the specific event as payload to the netdev event. At the moment the handler is done, this information is lost, so I can not access it in the ``_isr()``.  Attaching it to the netdev structure (or ``rail_t``) is not a good idea, because it would create a race condition if the next event comes in, while the older is still processed. Or do I miss something?

Reply:
Ok, I added an event queue, so the netdev->recv function can access the last received frame.